File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function(head) {
    let middle = head;
    let end = head;
    // while loop to iterate through the linked list
    while (end !== null && end.next !== null) {
        // checks if the (middle node) slow pointer is moved one step at a time 
        middle = middle.next;
        // checks if the (end node) fast pointer and its next node - end.next - moves up 2 nodes every step
        end = end.next.next;
    }
    return middle;

    // Time Complexity: O(n)
    // Space Complexity: O(1)
};

File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Create two pointers, slow and fast and set them to the head of the list
        slow = fast = head

        # Loop, while valid
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        # When fast hit the end of the list, slow will sit at the middle
        return slow